### ISSUE TITLE
Allow exceptions to be added to camelcase

### ISSUE THREAD

**Issue by smably** [eslint/eslint#10503]
**What rule do you want to change?**
`camelcase`

**Does this change cause the rule to produce more or fewer warnings?**
Fewer (if the `exceptions` option is used).

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option, something like:
```js
"camelcase": ['error', {
  "exceptions": [
    "UNSAFE_componentDidMount",
    "UNSAFE_componentWillReceiveProps",
    "UNSAFE_componentWillUpdate"
  ]
}]
```

**Please provide some example code that this change will affect:**

```js
import React from 'react';

class UnsafeComponent extends React.Component {
  UNSAFE_componentWillMount() {
    // deprecated lifecycle code
  }
  render() {
    return <div>FIXME</div>;
  }
}
```
[Demo link](https://eslint.org/demo/#eyJ0ZXh0IjoiaW1wb3J0IFJlYWN0IGZyb20gJ3JlYWN0JztcblxuY2xhc3MgVW5zYWZlQ29tcG9uZW50IGV4dGVuZHMgUmVhY3QuQ29tcG9uZW50IHtcbiAgVU5TQUZFX2NvbXBvbmVudFdpbGxNb3VudCgpIHtcbiAgICAvLyBkZXByZWNhdGVkIGxpZmVjeWNsZSBjb2RlXG4gIH1cbiAgcmVuZGVyKCkge1xuICAgIHJldHVybiA8ZGl2PkZJWE1FPC9kaXY+O1xuICB9XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo4LCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7ImpzeCI6dHJ1ZX19LCJydWxlcyI6eyJjYW1lbGNhc2UiOjJ9LCJlbnYiOnt9fX0=)

**What does the rule currently do for this code?**
Errors on `UNSAFE_componentWillMount`.

**What will the rule do after it's changed?**
Ignore `UNSAFE_componentWillMount` even though it's not camelCase.

-------------

As React deprecates its [legacy lifecycle methods](https://reactjs.org/docs/react-component.html#legacy-lifecycle-methods), we're going to see a proliferation of the new `UNSAFE_`-prefixed lifecycle methods. There's currently no good way to ignore these while enforcing `camelCase` elsewhere.

`id-match` is one possible workaround, but it results in an ugly and verbose rule that doesn't clearly describe its intent. Compare the proposed `camelcase` option above to the equivalent `id-match` rule:
```js
'id-match': [
  'error',
  '^(UNSAFE_componentDidMount|UNSAFE_componentWillReceiveProps|UNSAFE_componentWillUpdate|[a-z]+([A-Z][a-z]+)*)$'
]
```

This proposal was previously discussed in #9233 but didn't achieve consensus. I think the new option is worth revisiting in the context of React's new lifecycle methods, as it will be a much more frequent pain point for developers.

**Comment by platinumazure** [comment]
Hi @smably, thanks for creating a new issue.

I could get behind either a list of identifiers or a regular expression. (Maybe we could convert an identifier list into a regex behind the scenes?)

**Comment by sindresorhus** [comment]
I think it should support `Array<string|RegExp>`. Strings for the common-case, and RegExp for maximum power. In this case, `[/UNSAFE_component.*/]` would be much easier.

**Comment by AlicanC** [comment]
Can't "eslint-plugin-react" solve this issue by disabling `camelcase` for these, just like "eslint-plugin-prettier" disables style rules?

Also disabling `camelcase` for all usage doesn't solve this issue correctly. These should still be errors:
```javascript
const UNSAFE_componentDidMount = 13;

function UNSAFE_componentDidMount() {}

class MySomething extends NotReactComponent {
  UNSAFE_componentDidMount() {}
}
```

**Comment by platinumazure** [comment]
I'll champion this.

@eslint/eslint-team Can we review this one? It's pretty popular among the community, allowing the convenience of camelcase while also allowing users to list a small number of exceptions. Hopefully minimal implementation cost.

**Comment by nzakas** [comment]
I think it's a good idea. I would suggest changing `exceptions` to `allow` for better clarity, but otherwise, 👍 .

**Comment by hjylewis** [comment]
Is someone working on this? I would be happy to take a crack at it.

**Comment by platinumazure** [comment]
Hi @hjylewis, thanks for reaching out!

It looks like PR #10783 was independently created before this issue and is not quite following the proposal we eventually accepted here. I think I'll ping in that PR and see if they're willing to make the changes to match this issue; if they're busy/don't reply, then you could take a shot.

**Comment by platinumazure** [comment]
Is this issue resolved with #10783?

**Comment by platinumazure** [comment]
I'm going to close this as it seems it should be resolved with #10783. Please feel free to leave a comment if that's not the case. Thanks!
